### PR TITLE
Scripting: support CAN read and write

### PR DESCRIPTION
**Pull request by IamPete1**
Lots of nice generator improvements. All seems to work so long as there are no CAN messages before the driver is init, messages before the script comes up will result in a hard lockup. 

I suspect something to do with threading, not sure what yet.....

### CODE REVIEW

File: libraries/AP_Scripting/examples/CAN_read.lua [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  local id = frame:id()
+  id = id & 0xFFFFFFF
+
+  gcs:send_text(0,string.format("CAN msg from " .. tostring(id) .. ": %i, %i, %i, %i, %i, %i, %i, %i", frame:data(0), frame:data(1), frame:data(2), frame:data(3), frame:data(4), frame:data(5), frame:data(6), frame:data(7)))
+
+  return update, 1000 -- reschedules the loop
+
+end
+
+return update()
```
Comment (bugobliterator) [inline libraries/AP_Scripting/examples/CAN_read.lua:27]: newline

File: libraries/AP_Scripting/AP_Scripting_CANSensor.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    // handler for outgoing frames, using uint32
+    bool write_frame(AP_HAL::CANFrame &out_frame, const uint32_t timeout_us) {
+        return CANSensor::write_frame(out_frame, timeout_us);
+    };
+
+    // handler for incoming frames, add to buffer
+    void handle_frame(AP_HAL::CANFrame &frame) override { buffer.push(frame); }
+
+    // read a frame from the buffer
+    bool read_frame(AP_HAL::CANFrame &frame) { return buffer.pop(frame); }
```
Comment (bugobliterator) [inline libraries/AP_Scripting/AP_Scripting_CANSensor.h:37]: you need semaphore locking here, CANSensor runs on a thread.

**Review by bugobliterator — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/AP_Scripting_CANSensor.h (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_Scripting/AP_Scripting_CANSensor.h:37]: I'm using the thread safe ObjectBuffer, it does its own locking.

Comment (IamPete1) [reply libraries/AP_Scripting/AP_Scripting_CANSensor.h:39]: I assume that is sufficient anyway, its still locking up AP completely if there are messages before the script has run.

**Comment by IamPete1** [comment]
I have fixed the hard lockup by adding a 10ms delay in the CAN sensor thread. This includes a read / write example. It does a RGB fade using then UAVCAN `LightsCommand` message. Note that the light node must have a static ID.

**Comment by bugobliterator** [comment]
> I have fixed the hard lockup by adding a 10ms delay in the CAN sensor thread. This includes a read / write example. It does a RGB fade using then UAVCAN `LightsCommand` message. Note that the light node must have a static ID.

If you are transmitting UAVCAN message using this why not connect to UAVCAN library instead of creating a separate CANSensor?

**Comment by IamPete1** [comment]
> If you are transmitting UAVCAN message using this why not connect to AP_UAVCAN library instead of creating a separate CANSensor?

UAVCAN just makes for a convenient example with known working hardware, of course this is not the best way to UAVCAN. The use case is new CAN protocols or new messages for existing protocols.

File: libraries/AP_Scripting/generator/src/main.c [hunk 3]
```diff
@@ -1339,22 +1397,22 @@ void emit_userdata_field(const struct userdata *data, const struct userdata_fiel
         error(ERROR_INTERNAL, "Can't access a literal field");
         break;
       case TYPE_STRING:
-        fprintf(source, "            lua_pushstring(L, ud->%s);\n", field->name);
+        fprintf(source, "            lua_pushstring(L, ud->%s%s);\n", field->name, index_string);
         break;
       case TYPE_USERDATA:
-        error(ERROR_USERDATA, "Userdata does not currently support accss to userdata field's");
+        error(ERROR_USERDATA, "Userdata does not currently support acces to userdata field's");
```
Comment (davidbuzz) [inline libraries/AP_Scripting/generator/src/main.c:1403]: access

File: libraries/AP_Scripting/generator/src/main.c [hunk 4]
```diff
@@ -1339,22 +1397,22 @@ void emit_userdata_field(const struct userdata *data, const struct userdata_fiel
...
         break;
       case TYPE_AP_OBJECT: // FIXME: collapse the identical cases here, and use the type string function
-        error(ERROR_USERDATA, "AP_Object does not currently support accss to userdata field's");
+        error(ERROR_USERDATA, "AP_Object does not currently support acces to userdata field's");
```
Comment (davidbuzz) [inline libraries/AP_Scripting/generator/src/main.c:1406]: access here too.

File: libraries/AP_Scripting/generator/src/main.c (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_Scripting/generator/src/main.c]: fixed, serves me right for trying to correct a existing spelling error I guess...

**Review by bugobliterator — APPROVED** [review]

File: libraries/AP_CANManager/AP_CANSensor.cpp [hunk 5]
```diff
@@ -142,6 +142,7 @@ bool CANSensor::write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout_
 
 void CANSensor::loop()
 {
+    hal.scheduler->delay(10);
```
Comment (tridge) [inline libraries/AP_CANManager/AP_CANSensor.cpp:145]: we need to find why this is needed

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
@bugobliterator the delay(10) being needed worries me, also it doesn't enforce it using driver type 10, which means we could end up with two threads doing select on the same iface. That may not be safe

**Comment by tridge** [comment]
also a bit worried about thread priority

File: libraries/AP_CANManager/AP_CANSensor.cpp (hunk 5 above)
Comment (bugobliterator) [reply libraries/AP_CANManager/AP_CANSensor.cpp:145]: yes, I agree that we need to find why this is required. Although, this delay is only at the startup of the thread, so don't think it affects current operation.

**Comment by bugobliterator** [comment]
@tridge @IamPete1 I will try and find time to debug the lockup issue. Hopefully I can get to the bottom of this.

**Comment by IamPete1** [comment]
I have been able to fix the debug linker error by replacing the binding range value 

`AP_HAL::CANFrame::MaxDataLen`

with 

`int(ARRAY_SIZE(ud->data))`

Seems dumb that we would have to do that, but I'm not sure how else to fix.

This was the linking error, only with the debug build. 

`/ardupilot/build/sitl/libraries/AP_Scripting/lua_generated_bindings.cpp:260: undefined reference to AP_HAL::CANFrame::MaxDataLen`

File: libraries/AP_CANManager/AP_CANSensor.cpp (hunk 5 above)
Comment (IamPete1) [reply libraries/AP_CANManager/AP_CANSensor.cpp:145]: I have done a little more testing, it works fine with a 1ms delay too. If I comment the `handle_frame(frame);` it runs with no delay. If I change the new `handle_frame` function to be empty it does not run. So it is might be something to do with the virtual function. Its still not very clear. If I remove `int16_t res = _can_iface->receive(frame, time, flags);` and put `int16_t res = 1` it doesn't run.

**Comment by tridge** [comment]
@bugobliterator will have a look at the delay

File: libraries/AP_Scripting/AP_Scripting_CANSensor.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#pragma once
+
+#include <AP_CANManager/AP_CANSensor.h>
+
+#if HAL_MAX_CAN_PROTOCOL_DRIVERS
+class ScriptingCANBuffer;
+class ScriptingCANSensor : public CANSensor {
+public:
+
+    ScriptingCANSensor():CANSensor("Script", AP_CANManager::Driver_Type::Driver_Type_Scripting) {};
```
Comment (bugobliterator) [inline libraries/AP_Scripting/AP_Scripting_CANSensor.h:28]:
```suggestion
    ScriptingCANSensor():CANSensor("Script") {
        register_driver(AP_CANManager::Driver_Type::Driver_Type_Scripting);
    }
```

@IamPete1  please rebase on top of this PR https://github.com/ArduPilot/ardupilot/pull/18089 . Add above changes and you should be able to remove the added delay.

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Requires this PR https://github.com/ArduPilot/ardupilot/pull/18089 and adding fix as suggested. Also need to remove the added delay added because of the bug the PR fixes.

**Comment by IamPete1** [comment]
```
Binary           text             data         bss            total
---------------  ---------------  -----------  -------------  ---------------
antennatracker   1700 (+0.1308%)  0 (0.0000%)  4 (+0.0015%)   1704 (+0.1091%)
arducopter       1708 (+0.1016%)  0 (0.0000%)  4 (+0.0015%)   1712 (+0.0881%)
arducopter-heli  1692 (+0.1020%)  0 (0.0000%)  -4 (-0.0015%)  1688 (+0.0879%)
ardurover        1692 (+0.1124%)  0 (0.0000%)  -4 (-0.0015%)  1688 (+0.0955%)
arduplane        1696 (+0.1008%)  0 (0.0000%)  0 (0.0000%)    1696 (+0.0872%)
blimp            1708 (+0.1394%)  0 (0.0000%)  -4 (-0.0015%)  1704 (+0.1145%)
ardusub          1732 (+0.1154%)  0 (0.0000%)  -4 (-0.0015%)  1728 (+0.0980%)
```

**Review by tridge — APPROVED** [review]

**Pull request merged**